nilearn/image/resampling.py

```diff
@@ -657,42 +545,235 @@ def resample_img(
+                # Before scipy 0.18, ndimage.affine_transform was applying
+                # a different logic to the offset for diagonal affine
                 b = np.dot(linalg.inv(A), b)
             A = np.diag(A)
-        # Iterate over a set of 3D volumes, as the interpolation problem is
-        # separable in the extra dimensions. This reduces the
-        # computational cost
+
+        # Iterate over a set of 3D volumes, as the interpolation problem
+        # is separable in the extra dimensions.
+        # This reduces the computational cost
         for ind in np.ndindex(*other_shape):
             _resample_one_img(
                 data[all_img + ind],
                 A,
                 b,
                 target_shape,
-                interpolation_order,
+                _interpolation_order(interpolation),
                 out=resampled_data[all_img + ind],
                 copy=not input_img_is_string,
                 fill_value=fill_value,
             )
 
-    if clip:
-        # force resampled data to have a range contained in the original data
-        # preventing ringing artefact
-        # We need to add zero as a value considered for clipping, as it
-        # appears in padding images.
-        vmin = min(data.min(), 0)
-        vmax = max(data.max(), 0)
-        resampled_data.clip(vmin, vmax, out=resampled_data)
+    resampled_data = _clip_img(data, resampled_data, clip)
 
     return new_img_like(img, resampled_data, target_affine)
 
 
+def _get_resampled_datatype(interpolation, data, transform_affine):
+    """Cast unsupported data types to closest support dtype."""
+    if interpolation == "continuous" and data.dtype.kind == "i":
+        aux = data.dtype.name.replace("int", "float")
+        aux = aux.replace("ufloat", "float").replace("floatc", "float")
+        if aux in ["float8", "float16"]:
+            aux = "float32"
+        warnings.warn(f"Casting data from {data.dtype.name} to {aux}")
+        resampled_data_dtype = np.dtype(aux)
+    else:
+        resampled_data_dtype = data.dtype
+
+    # Since the release of 0.17, resampling nifti images have some issues
+    # when affine is passed as 1D array
+    # and if data is of non-native endianness.
+    # See issue https://github.com/nilearn/nilearn/issues/1445.
+    # If affine is passed as 1D, scipy uses _nd_image.zoom_shift rather
+    # than _geometric_transform (2D) where _geometric_transform is able
+    # to swap byte order in scipy later than 0.15 for nonnative endianness.
+
+    # We convert to 'native' order to not have any issues either with
+    # 'little' or 'big' endian data dtypes (non-native endians).
+    A, _ = to_matrix_vector(transform_affine)
+    if len(A.shape) == 1 and not resampled_data_dtype.isnative:
+        resampled_data_dtype = resampled_data_dtype.newbyteorder("N")
+
+    return resampled_data_dtype
+
+
+def _get_bounding_box(target_affine, target_shape, affine, data):
```

Comment:
maybe this could be called something like `_check_target_affine_and_shape` -- functions in nilearn & scikit-learn that validate inputs and transform them or load defaults if necessary are often called `_check` something. and the docstring could state that this is a helper for `resample_img` to infer the missing bits of target affine and shape when they are not fully specified. side note -- it's ok to keep docstrings of private functions short, for example here as the parameters are those of resample_img you could skip it. the `Returns` section would actually be more informative